lib/SPIRVProducerPass.cpp

```diff
@@ -5915,3 +5878,87 @@ bool SPIRVProducerPass::CalledWithCoherentResource(Argument &Arg) {
   // No coherent resource variables encountered.
   return false;
 }
+
+void SPIRVProducerPass::PopulateStructuredCFGMaps(Module &module) {
+  // First, track loop merges and continues.
+  DenseSet<BasicBlock *> LoopMergesAndContinues;
+  for (auto &F : module) {
+    if (F.isDeclaration())
+      continue;
+
+    DominatorTree &DT = getAnalysis<DominatorTreeWrapperPass>(F).getDomTree();
+    const LoopInfo &LI = getAnalysis<LoopInfoWrapperPass>(F).getLoopInfo();
+    std::deque<BasicBlock *> order;
+    DenseSet<BasicBlock *> visited;
+    clspv::ComputeStructuredOrder(&*F.begin(), &DT, LI, &order, &visited);
+
+    for (auto BB : order) {
+      auto terminator = BB->getTerminator();
+      auto branch = dyn_cast<BranchInst>(terminator);
+      auto L = LI.getLoopFor(BB);
+      if (LI.isLoopHeader(BB)) {
+        BasicBlock *ContinueBB = nullptr;
+        BasicBlock *MergeBB = nullptr;
+
+        MergeBB = L->getExitBlock();
+        if (!MergeBB) {
+          // StructurizeCFG pass converts CFG into triangle shape and the cfg
+          // has regions with single entry/exit. As a result, loop should not
+          // have multiple exits.
+          llvm_unreachable("Loop has multiple exits???");
+        }
+
+        if (L->isLoopLatch(BB)) {
+          ContinueBB = BB;
+        } else {
+          // From SPIR-V spec 2.11, Continue Target must dominate that back-edge
+          // block.
+          BasicBlock *Header = L->getHeader();
+          BasicBlock *Latch = L->getLoopLatch();
+          for (auto *loop_block : L->blocks()) {
+            if (loop_block == Header) {
+              continue;
+            }
+
+            // Check whether block dominates block with back-edge.
+            if (DT.dominates(loop_block, Latch)) {
+              ContinueBB = loop_block;
```

Comment:
Also need the backedge block to post-dominate the continue target.
How do we know that will be satisfied as well?

Reply:
Hmm. Seems like this is a consequence of the StructurizeCFG algorithm?
Needs a comment at least.